scanner.go

```diff
@@ -359,3 +378,63 @@ func (s *scanner) closeRegionScanner() {
 	}
 	s.curRegionScannerID = noScannerID
 }
+
+// renews a scanner by resending scan request with renew = true
+func (s *scanner) renew() error {
+	if err := s.rpc.Context().Err(); err != nil {
+		return err
+	} else {
+		rpc, err := hrpc.NewScanRange(s.rpc.Context(),
+			s.rpc.Table(),
+			s.startRow,
+			nil,
+			hrpc.ScannerID(s.curRegionScannerID),
+			hrpc.NumberOfRows(s.rpc.NumberOfRows()),
+			hrpc.Priority(s.rpc.Priority()),
+			hrpc.RenewalScan(),
+		)
+		if err != nil {
+			return err
+		}
+		res, err := s.SendRPC(rpc)
+		if err != nil {
+			return err
+		}
+		resp, ok := res.(*pb.ScanResponse)
+		if !ok {
+			return errors.New("got non-ScanResponse for scan request")
+		}
+		// save new scanner id
+		s.curRegionScannerID = resp.GetScannerId()
+		return nil
+	}
+}
+
+// starts the renew timer, after it fires, renewScanner is called
+func (s *scanner) startRenewTimer() {
+	if s.renewTimer != nil {
+		panic("Renew timer was not stopped before starting again")
+	}
+	if s.rpc.RenewInterval() > 0 {
+		s.renewTimer = time.AfterFunc(s.rpc.RenewInterval(), s.renewScanner)
```

Comment:
A different way to do this would be to start a goroutine that sleeps on a ticker and context.

```
renewCtx, cancel := context.WithCancel(s.Context())
s.renewCancel = cancel
go s.renewLoop(renewCtx)
```

And `renewLoop` would look like:
```
func (s *scanner) renewLoop() {
    t := time.Ticker(s.rpc.RenewInterval()
    defer t.Stop()
    for {
        select {
        case <-t.C:
             if err := s.renew(); err != nil {
                    return
              }
        case <-ctx.Done():
            return
        }
    }
}
```

The advantage here is that the renewLoop routine will clean itself up immediately after the context is canceled and doesn't have to wait for the timer to go off to notice the context is done. And I think it's slightly simpler to have a loop rather than `renewScanner()` needing to start a new renew timer.

Reply:
Another advantage here is that it provides a place for us to add a metric to track the number of live renewals. It can be done in a follow on change, but we will want to add a metric that increments when this function starts and decrements when it exits. That way we can track if we are leaking scanners.

Reply:
Thanks, this definitely cleans up the code and makes it easier to add metrics in the future. I've implemented this, feel free to take a look!